File: kinv/adapters/kicad/csv.py
"""RFC 4180 CSV reader.

Written rather than taken from the ``csv`` module because it has to agree with
the TypeScript reader to the character: the reference-designator column is
full of embedded commas (``"C1001,C2004,C6001-C6004"``), and the file may
arrive with CRLF, LF, or a byte-order mark.
"""

from __future__ import annotations


def parse_csv(text: str, delimiter: str = ",") -> list[list[str]]:
    if text.startswith("﻿"):
        text = text[1:]
    rows: list[list[str]] = []
    row: list[str] = []
    field: list[str] = []
    quoted = False
    saw_field = False
    i = 0
    n = len(text)

    def end_field() -> None:
        nonlocal saw_field
        row.append("".join(field))
        field.clear()
        saw_field = False

    def end_row() -> None:
        nonlocal row
        end_field()
        rows.append(row)
        row = []

    while i < n:
        ch = text[i]

        if quoted:
            if ch == '"':
                if i + 1 < n and text[i + 1] == '"':
                    field.append('"')
                    i += 2
                    continue
                quoted = False
                i += 1
                continue
            field.append(ch)
            i += 1
            continue

        if ch == '"' and not saw_field:
            quoted = True
            saw_field = True
            i += 1
            continue
        if ch == delimiter:
            end_field()
            i += 1
            continue
        if ch == "\r":
            i += 2 if i + 1 < n and text[i + 1] == "\n" else 1
            end_row()
            continue
        if ch == "\n":
            i += 1
            end_row()
            continue

        field.append(ch)
        saw_field = True
        i += 1

    # A trailing newline must not produce a phantom row.
    if field or row:
        end_row()

    return rows


def parse_csv_records(text: str, delimiter: str = ",") -> list[dict[str, str]]:
    """Rows keyed by header name, with the header row consumed."""
    rows = parse_csv(text, delimiter)
    if not rows:
        return []
    header = rows.pop(0)
    records = []
    for row in rows:
        if not any(cell.strip() != "" for cell in row):
            continue
        record: dict[str, str] = {}
        for index, name in enumerate(header):
            record[name] = row[index] if index < len(row) else ""
        records.append(record)
    return records

File: src/adapters/kicad/csv.ts
/**
 * RFC 4180 CSV reader.
 *
 * Written rather than pulled in because the reference-designator column is full
 * of embedded commas (`"C1001,C2004,C6001-C6004"`), so a naive split destroys
 * the data — and because the file may arrive with CRLF, LF, or a BOM.
 */
export function parseCsv(input: string, delimiter = ","): string[][] {
  const text = input.charCodeAt(0) === 0xfeff ? input.slice(1) : input;
  const rows: string[][] = [];
  let row: string[] = [];
  let field = "";
  let quoted = false;
  let i = 0;
  let sawField = false;

  const endField = (): void => {
    row.push(field);
    field = "";
    sawField = false;
  };
  const endRow = (): void => {
    endField();
    rows.push(row);
    row = [];
  };

  while (i < text.length) {
    const ch = text[i] as string;

    if (quoted) {
      if (ch === '"') {
        if (text[i + 1] === '"') {
          field += '"';
          i += 2;
          continue;
        }
        quoted = false;
        i += 1;
        continue;
      }
      field += ch;
      i += 1;
      continue;
    }

    if (ch === '"' && !sawField) {
      quoted = true;
      sawField = true;
      i += 1;
      continue;
    }
    if (ch === delimiter) {
      endField();
      i += 1;
      continue;
    }
    if (ch === "\r") {
      i += text[i + 1] === "\n" ? 2 : 1;
      endRow();
      continue;
    }
    if (ch === "\n") {
      i += 1;
      endRow();
      continue;
    }

    field += ch;
    sawField = true;
    i += 1;
  }

  // A trailing newline must not produce a phantom row.
  if (field !== "" || row.length > 0) endRow();

  return rows;
}

/** Rows keyed by header name, with the header row consumed. */
export function parseCsvRecords(input: string, delimiter = ","): Record<string, string>[] {
  const rows = parseCsv(input, delimiter);
  const header = rows.shift();
  if (!header) return [];
  return rows
    .filter((r) => r.some((cell) => cell.trim() !== ""))
    .map((r) => {
      const record: Record<string, string> = {};
      header.forEach((name, idx) => {
        record[name] = r[idx] ?? "";
      });
      return record;
    });
}
